### PR TITLE
Api keys route implemented

### PR DESCRIPTION
**Pull request by AjmalAhamedKodakkat**

### CODE REVIEW

File: binance-smart-chain-rpc/src/constructors/ethereum/compiler.js [hunk 1]
```diff
@@ -1,154 +1,184 @@
-const path = require('path')
```
Comment (srslafazan) [inline binance-smart-chain-rpc/src/constructors/ethereum/compiler.js:1]: We want to separate PRs for different features, so this is best to go in a different commit and PR. We'll review them together this time.

File: binance-smart-chain-rpc/src/constructors/ethereum/compiler.js [hunk 2]
```diff
@@ -1,154 +1,184 @@
...
-  contractJson,
-  sender,
-  privateKey,
-  sendOptions,
-  deployArgs = [],
+ contractJson,
+ sender,
+ privateKey,
+ sendOptions,
+ deployArgs = [],
```
Comment (srslafazan) [inline binance-smart-chain-rpc/src/constructors/ethereum/compiler.js:88]: Let's install eslint when we are bored or if we have some gap in tasks. I've skipped it as a single developer, but it's better practice to have more uniform code formatting.

File: binance-smart-chain-rpc/src/constructors/express/routes/api/v1/tokens.js [hunk 3]
```diff
@@ -1,43 +1,54 @@
...
-      return res.status(500).json({ errors: [e] })
+module.exports = ({
+ models,
+ ethereum: { web3, buildContract, compiler, abi },
+ ...context
+}) => {
+ 
+ router.post(
+  "/",
+  [body("type").trim().isIn(["APIS_ERC20", "APIS_ERC721"])],
```
Comment (srslafazan) [inline binance-smart-chain-rpc/src/constructors/express/routes/api/v1/tokens.js:15]: awesome

File: docker-compose.yml [hunk 4]
```diff
@@ -17,7 +17,7 @@ services:
       - ./gateway/src:/app
       - ./share:/app/share
       - ./share/models:/app/share/models
-      - ./share/migrations:/app/share/sequelize/migrations
+      - ./share/sequelize/migrations/client:/app/share/sequelize/migrations
```
Comment (srslafazan) [inline docker-compose.yml:20]: 👍

File: gateway/src/constructors/express/express.js [hunk 5]
```diff
@@ -4,13 +4,13 @@ const cors = require('cors')
...
-module.exports = async ({ io }) => {
+module.exports = async ({ io,...context }) => {
   const app = express()
   app.disable('x-powered-by')
   app.use(morgan('combined'))
   app.use(bodyParser.json({ limit: '50mb', extended: true }))
   app.use(bodyParser.urlencoded({ limit: '50mb', extended: true }))
   app.use(cors())
-  app.use('/', require('./routes')({}))
+  app.use('/', require('./routes')({...context}))
```
Comment (srslafazan) [inline gateway/src/constructors/express/express.js:14]: we probably don't need to copy this object here

File: gateway/src/constructors/express/routes/index.js [hunk 6]
```diff
@@ -1,30 +1,45 @@
...
-    console.log('TODO - authenticate...')
-    next()
-  })
-
-  if (process.env.NODE_ENV === 'development') {
-    router.use('/api', cors(), APIRouter)
-  } else {
-    router.use('/api', APIRouter)
+ router.use("/api", async (req, res, next) => {
+  console.log("TODO - authenticate....");
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/index.js:11]: we can remove this TODO

File: gateway/src/constructors/express/routes/index.js [hunk 7]
```diff
@@ -1,30 +1,45 @@
...
-const router = require('express').Router()
+const path = require("path");
+const cors = require("cors");
+const router = require("express").Router();
 
 module.exports = (context = {}) => {
-  const APIRouter = require('./api')(context)
+ const { sequelize, models } = context;
+ const APIRouter = require("./api")(context);
+ const KEYRouter = require("./key")(context);
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/index.js:8]: Let's have a more generic strategy for this.

- `.../api/*`
- `.../public/*`

ex: 
- `.../public/v1/apikeys/*`

File: gateway/src/constructors/express/routes/index.js [hunk 8]
```diff
@@ -1,30 +1,45 @@
...
-    router.use('/api', APIRouter)
+ router.use("/api", async (req, res, next) => {
+  console.log("TODO - authenticate....");
+  try {
+   const { apiKey = "*" } = req.query;
+   const Op = sequelize.Op;
+   const result = await models.APIKey.findAll({ where: { api_key: apiKey } });
+   if (result.length == 0) {
+    return res.status(401).json({ error: "API KEY NOT VALID" });
+   }
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/index.js:18]: We should use `APIKey.findOne(...)`.

If the `result === null`, you know the key record was not found.

File: gateway/src/constructors/express/routes/index.js [hunk 9]
```diff
@@ -1,30 +1,45 @@
...
-  } else {
-    router.use('/api', APIRouter)
+ router.use("/api", async (req, res, next) => {
+  console.log("TODO - authenticate....");
+  try {
+   const { apiKey = "*" } = req.query;
+   const Op = sequelize.Op;
+   const result = await models.APIKey.findAll({ where: { api_key: apiKey } });
+   if (result.length == 0) {
+    return res.status(401).json({ error: "API KEY NOT VALID" });
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/index.js:17]: A better error message might be, `INVALID API KEY`.

File: gateway/src/constructors/express/routes/index.js [hunk 10]
```diff
@@ -1,30 +1,45 @@
...
-  router.use('/api', (req, res, next) => {
-    console.log('TODO - authenticate...')
-    next()
-  })
-
-  if (process.env.NODE_ENV === 'development') {
-    router.use('/api', cors(), APIRouter)
-  } else {
-    router.use('/api', APIRouter)
+ router.use("/api", async (req, res, next) => {
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/index.js:10]: We may use `express-validator` to respond to a missing api key parameter.

File: gateway/src/constructors/express/routes/index.js [hunk 11]
```diff
@@ -1,30 +1,45 @@
...
+    return res.status(401).json({ error: "API KEY NOT VALID" });
+   }
+   next();
+  } catch (e) {
+   console.error(e);
+   return res.status(401).json({ error: [e] });
   }
+ });
+
+ router.use("/model/api/key", KEYRouter)
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/index.js:26]: Something like: `/public/v1/apikeys` will work. It's standard practice with REST routes to have version namespaces and to use plural path tokens.

File: gateway/src/constructors/express/routes/key/index.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
+const axios = require('axios')
+const router = require('express').Router()
+
+
+module.exports = (context) => {
+ const { sequelize, models } = context;
+  router.post('/', async (req, res, next) => {
+    try {
+   const { name = "*", email = "*" } = req.query;
+   console.log("model.....",models)
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/key/index.js:10]: We may remove these logs -- or use `debug` statements.

File: gateway/src/constructors/express/routes/key/index.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
+const axios = require('axios')
+const router = require('express').Router()
+
+
+module.exports = (context) => {
+ const { sequelize, models } = context;
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/key/index.js:6]: If we prefer, these may be pulled straight from the context (since we know they're always there):

```
module.exports = ({ sequelize, models, ...context }) => {
```

File: gateway/src/constructors/express/routes/key/index.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+  router.post('/', async (req, res, next) => {
+    try {
+   const { name = "*", email = "*" } = req.query;
+   console.log("model.....",models)
+   
+   const Op = sequelize.Op;
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/key/index.js:12]: Looks like this `Op` is unused

**Review by srslafazan — CHANGES_REQUESTED** [review]
Great first commit! 🥇 Some changes to be made in the comments.

File: binance-smart-chain-rpc/src/constructors/ethereum/compiler.js (hunk 1 above)
Comment (AjmalAhamedKodakkat) [reply binance-smart-chain-rpc/src/constructors/ethereum/compiler.js:1]: Sure, I will make a separate PR from my forked repo to the master . I was working parallelly on bsc and api-keys , that's why it came with branch commit. I will do a separate PR once I finish the bsc setup for token launch . We may keep the bsc master and api-keys branch seperately. Once finishing this , I will do the changes in api-key branch

File: binance-smart-chain-rpc/src/constructors/ethereum/compiler.js (hunk 2 above)
Comment (AjmalAhamedKodakkat) [reply binance-smart-chain-rpc/src/constructors/ethereum/compiler.js:88]: Cool, Right now I am using prettier with 1 space for formatting code  . Would like to know to use eslint

File: gateway/src/constructors/express/routes/index.js (hunk 6 above)
Comment (AjmalAhamedKodakkat) [reply gateway/src/constructors/express/routes/index.js:11]: sure will do that

File: gateway/src/constructors/express/routes/index.js (hunk 7 above)
Comment (AjmalAhamedKodakkat) [reply gateway/src/constructors/express/routes/index.js:8]: ok

File: gateway/src/constructors/express/routes/index.js (hunk 8 above)
Comment (AjmalAhamedKodakkat) [reply gateway/src/constructors/express/routes/index.js:18]: ok, i will try with that

**Comment by AjmalAhamedKodakkat** [comment]
Overall changes according to the review will be done asap after finishing the bsc token launch PR. and then afterward i will move to row encryptions. Thanks for the review

File: binance-smart-chain-rpc/src/constructors/ethereum/compiler.js (hunk 2 above)
Comment (srslafazan) [reply binance-smart-chain-rpc/src/constructors/ethereum/compiler.js:87]: Oh! I see -- let's change prettier to 2 spaces

File: gateway/src/constructors/express/express.js [hunk 15]
```diff
@@ -4,13 +4,13 @@ const cors = require('cors')
...
-module.exports = async ({ io }) => {
+module.exports = async ({ io,...context }) => {
   const app = express()
   app.disable('x-powered-by')
   app.use(morgan('combined'))
   app.use(bodyParser.json({ limit: '50mb', extended: true }))
   app.use(bodyParser.urlencoded({ limit: '50mb', extended: true }))
   app.use(cors())
-  app.use('/', require('./routes')({}))
+  app.use('/', require('./routes')({context}))
```
Comment (srslafazan) [inline gateway/src/constructors/express/express.js:14]: I think we need to pass this without the braces: `app.use('/', require('./routes')(context))`

File: gateway/src/constructors/express/routes/index.js [hunk 16]
```diff
@@ -1,30 +1,50 @@
...
-const router = require('express').Router()
+const cors = require("cors");
+const router = require("express").Router();
+const {  query, validationResult } = require("express-validator");
 
-module.exports = (context = {}) => {
-  const APIRouter = require('./api')(context)
+module.exports = ({sequelize, models,...context }) => {
+ const APIRouter = require("./api")(context);
+ const PUBLICRouter = require("./public")(context);
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/index.js:7]: Perhaps we should call it `PublicRouter`, with no need to have full capitals for PUBLIC

File: share/sequelize/seeders/20200920195019-EthereumContract.js [hunk 17]
```diff
@@ -1,68 +0,0 @@
-'use strict';
```
Comment (srslafazan) [inline share/sequelize/seeders/20200920195019-EthereumContract.js:1]: Did we intend to delete this file?

File: gateway/src/constructors/express/routes/public/v1/index.js [hunk 18]
```diff
@@ -0,0 +1,8 @@
+
+const router = require('express').Router()
+
+
+module.exports = ({...context}) => {
+ router.use('/apiKeys', require('./apiKeys')(context))
```
Comment (srslafazan) [inline gateway/src/constructors/express/routes/public/v1/index.js:6]: Let's use lowercase, `/apikeys`

**Review by srslafazan — CHANGES_REQUESTED** [review]
Almost there! Just minor comments remaining

File: gateway/src/constructors/express/express.js (hunk 15 above)
Comment (AjmalAhamedKodakkat) [reply gateway/src/constructors/express/express.js:14]: the code is not complete, i  just pushed it bcz , my lap just had some problem, so just pushed the current status to make sure  that i dont loose the work done...will do the changes mentioned in the review

File: share/sequelize/seeders/20200920195019-EthereumContract.js (hunk 17 above)
Comment (AjmalAhamedKodakkat) [reply share/sequelize/seeders/20200920195019-EthereumContract.js:1]: yes, bcz it was obstructing the migrations . Do i need to replace that?

Comment (AjmalAhamedKodakkat) [reply share/sequelize/seeders/20200920195019-EthereumContract.js:1]: i have replaced it

**Review by srslafazan — APPROVED** [review]
LGTM!
